net: sntp: Add "simple" one-shot API to query SNTP server

pfalcon opened the pull request:
Many real-world tasks require absolute time on devices which otherwise lack RTC. A realistic way then to use SNTP to get the current time. There should be a way to do that quickly and easily. Introduced in this PR. (This PR also includes patches from #15582 & #15600 as dependencies)

pfalcon committed:
net: socketutils: Add utils to manipulate network address strings

Two utils to manipulate addresses in format "addr[:port]". I.e.,
network address (domain name or numeric), optionally followed by
port number:

* net_addr_str_find_port(), to return pointer to port number
substring (or NULL if not present).
* net_getaddrinfo_addr_str(), which is effectively getaddrinfo()
wrapper taking a "addr[:port]" string as a parameter.

The header file is named socketutils.h to emphasize that these
utility functions are implemented on top of BSD Sockets API
(and other POSIX/ANSI C functions), and thus portable to other
POSIX systems (e.g., Linux), so can be used in apps testing
POSIX compatibility. More utility functions (beyond address
manipulation) can be added later.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: sntp: Add convenience API for one-shot SNTP query

sntp_simple() function queries the server (passed as "addr[:port]"
string). It wraps calls to a number of other functions, and may be
useful to write simple, concise apps needing the absolute time.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
@d3zd3z, @galak: GoogleIoT SNTP support I'm working on depends on this patch.

jukkar commented on include/net/netutils.h:
```diff
@@ -0,0 +1,35 @@
+/*
```
We have already net_ip.h which contains various helper functions, so there is no need to create a new header file for just these.

pfalcon replied:
> We have already net_ip.h which contains various helper functions, so there is no need to create a new header file for just these.

So helper functions are for Zephyr network stack. These are application-level helpers on top of POSIX API.

jukkar replied:
Why not call this `utils.h` as the net is redundant info here as the file is in net directory anyway. We have other net_ prefixed header files in the net directory but it feels unnecessary now, unfortunately we cannot really change that at this point.

pfalcon replied:
> Why not call this utils.h as the net is redundant info here as the file is in net directory anyway.

As I mentioned, the idea was to make it have more selective name than just often-overused "utils", in particular, because of the idea to use it e.g. in POSIX (Linux) app builds, where "net/" prefix might be there or not.

Calling it just net/utils.h might work, but based on the previous discussion, it seems it's worth emphasizing that it's socket API level utils, so let me try socketutils.h and see if that works. (Also considered sockutils.h, but that may be confused with utils for SOCKS protocol for example).

jukkar commented on subsys/net/lib/utils/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
```
We have network utilities already in `subsys/net/ip/utils.c` file, no need to create a new file for these.

pfalcon replied:
> We have network utilities already in subsys/net/ip/utils.c file, no need to create a new file for these.

In subsys/net/ip/utils.c, we have internal utilities for Zephyr IP stack. This PR introduces generic, POSIX-compliant helper functions, which in particular intended to be portable and buildable with other POSIX systems. To elaborate, I initially thought of putting them under `lib/netutils/`, but as a compromise put under existing `subsys/net/lib/`. If their intentions there are not clear, we can move them under lib/.

jukkar commented on subsys/net/lib/utils/addr_utils.c:
```diff
@@ -0,0 +1,53 @@
+
+#include <net/socket.h>
+
+const char *net_addr_str_find_port(const char *addr_str)
```
This looks like it is doing similar things as what `net_ipaddr_parse()` is doing. Could you utilize that function here?

pfalcon replied:
> This looks like it is doing similar things as what net_ipaddr_parse() is doing.

Yes, I recalled that we should have a func like that and did my homework on that. So, first of all, net_ipaddr_parse() would be similar to net_getaddrinfo_addr_str(), not net_addr_str_find_port(). But it's not, because net_ipaddr_parse() only handles numeric addresses, whereas newly introduced net_getaddrinfo_addr_str() can handle domain names too (and implications of that, like the fact that domain names may resolve to multiple underlying addresses).

I agree that such perceived duplication is not ideal, but as explained, I did my homework on it. My idea of dealing with that is along the lines of my previous replies - treat existing functions like net_ipaddr_parse() as IP-stack level (I believe it's used (only?) to parse DNS server address, which has to be done on the stack level). Whereas idea here is to introduce a set of more generic application-level functions, implemented on top of POSIX API, and thus fully portable, e.g. to Linux too.

Another (parallel) outlook may be to treat functions like net_ipaddr_parse() as limited, but optimized functions, whereas net_getaddrinfo_addr_str() is generic, but has higher overhead. For example, I wouldn't propose to replace call to net_ipaddr_parse() in DNS code to net_getaddrinfo_addr_str() - it doesn't make sense to bring up full overhead of getaddrinfo(). (It might be possible to make net_ipaddr_parse() call newly introduced net_addr_str_find_port(), but looking at net_ipaddr_parse(), I'm not sure that'll give much benefit there).

Hope this explains in detail my outlook on the matter.

jukkar replied:
I propose following:
* Moving generic networking related helpers to `net/utils.h` from `net/net_ip.h`. The latter is not a perfect place for utility functions anyway.
* Modifying `net_ipaddr_parse()` to support DNS resolving if the input string is not a IP address.

pfalcon replied:
> Moving generic networking related helpers to net/utils.h from net/net_ip.h.

Minor, but I named it net/netutils.h to have better namespacing. Less minor, that I'd really like to keep it POSIX API based (and buildable). One of the usecases is pulling those files (netutils.h and the implementation) to another POSIX-compatible system and using there.

> Modifying net_ipaddr_parse() to support DNS resolving if the input string is not a IP address.

But why? Again, in the current codebase, only DNS initialization uses it (well, log_backend_net.c also calls it, but that can be considered just the same system-level). And DNS initialization requires a numeric address, so supporting domain names there is useless. And you can't easily modify net_ipaddr_parse() to deal with domain names, because then it needs to be modified to handle case of multiple IP addresses per domain name (e.g. IPv4 vs IPv6 addresses, or just multiple addresses at all). And if you don't support multiple return addresses, it remains a toy function, not suitable for all possible real-world uses, but which however was bloated up beyond its real usecase.

pfalcon commented:
@rlubos: This PR is what I was talking about previously re: API design. Once we have a set of good, generic, composable API primitives (*), we can provide whatever convenience APIs on top of them as needed (or even wished). For example, here the usecase and requirement is "Suppose you have a POSIX app, which of course expect a proper absolute time to be available, and you want to port it to Zephyr. It should be possible to obtain the absolute time with ~just one line (function call) (including no extra legwork on preparing params to this function call)".

* SNTP API is actually not such, and needs further work, e.g. currently it's synchronously only, which is not adequate.

vanti commented on include/net/sntp.h:
```diff
@@ -82,6 +82,20 @@ int sntp_query(struct sntp_ctx *ctx, u32_t timeout,
+ * @param timeout Query timeout
+ * @param time Timestamp including integer and fractional seconds since
+ * 1 Jan 1970 (output).
+ */
```
It'd be good to document the return value of this function.

pfalcon replied:
> It'd be good to document the return value of this function.

Fixed, thanks.

vanti commented on include/net/sntp.h:
```diff
@@ -82,6 +82,20 @@ int sntp_query(struct sntp_ctx *ctx, u32_t timeout,
+ * @brief Convenience function to query SNTP in one-shot fashion
+ *
+ * Convenience wrapper which calls getaddrinfo(), sntp_init(),
+ * sntp_request(), and sntp_close().
```
I think 'sntp_request()' should be 'sntp_query()'

pfalcon replied:
> I think 'sntp_request()' should be 'sntp_query()'

Good catch, fixed.

pfalcon commented:
@PiotrZierhoffer, @tgorochowik: This is an example of paving the road for "wider compliant POSIX env" setup. For me, this is intermediate step, it should culminate in having a config option like CONFIG_GIMME_REAL_TIME, and then when application start up, gettimeofday() returning proper POSIX compliant timestamp.

pfalcon commented on subsys/net/lib/utils/addr_utils.c:
```diff
@@ -0,0 +1,60 @@
+			     struct addrinfo **res)
+{
+	const char *port = net_addr_str_find_port(addr_str);
+	char host[64];
```
> NI_MAXHOST would be better here

Note that NI_MAXHOST is not in POSIX:
http://pubs.opengroup.org/cgi/susv4search.cgi?KEYWORDS=NI_MAXHOST&SUBSTRING=substring&CONTEXT=

Let me hack it up, but we'd better get the basics right, then add extensions properly.

jukkar replied:
> Note that NI_MAXHOST is not in POSIX:

I am not claiming it is in Posix. I am just saying that having magic value here is bad, thus I proposed this symbol.

pfalcon committed:
net: socket: Define NI_MAXHOST for compatibility

Not in POSIX. Linux man getnameinfo says about it:

"In order to assist the programmer in choosing reasonable sizes for
the supplied buffers, <netdb.h> defines the constants

           #define NI_MAXHOST      1025
           #define NI_MAXSERV      32

Since glibc 2.8, these definitions are exposed only if suitable
feature test macros are defined, namely: _GNU_SOURCE, _DEFAULT_SOURCE
(since glibc 2.19), or (in glibc versions up to and including 2.19)
_BSD_SOURCE or _SVID_SOURCE."

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented on subsys/net/lib/utils/addr_utils.c:
```diff
@@ -0,0 +1,60 @@
+			return -1;
+		}
+		strncpy(host, addr_str, host_len + 1);
+		host[host_len] = 0;
```
> Better host[host_len] = '\0'; as typically we write the terminating null like that.

It's so 1980ies, but if we typically write it like that, fixed.

pfalcon commented on subsys/net/lib/utils/addr_utils.c:
```diff
@@ -0,0 +1,60 @@
+ */
+#ifdef CONFIG_NET_SOCKETS_POSIX_NAMES
+
+const char *net_addr_str_find_port(const char *addr_str)
```
> Why do we expose this function to user application, does it really need it? Could we just make it static?

Yes, the idea was to expose both functions to user apps, to give them the flexibility. For example, if a user apps needs address only in numeric format, it can use net_addr_str_find_port() to parse it into addr vs port substrings, and then use inet_pton(), and avoid dependency on getaddrinfo().

jukkar replied:
I think we should keep this as static, it can be made global if really needed by someone. We should try to minimize the interfaces that we expose and this function is just too low level that application programmer would even find it in the API listing.

pfalcon replied:
> We should try to minimize the interfaces that we expose and this function is just too low level that application programmer would even find it in the API listing.

We should (finally) provide good and stable API set. And good API design exposes "atomic" functions, which can be combined by user to achieve any desired effect, while also providing higher-level convenience functions for commonly desired effect. That's exactly how this API is laid out. The API which offers only higher-level functionality and locks out user from using it, or require rewriting it completely to adjust some behavior, is well, bad API.

I don't see why all this concern about making that function static. Making it static is inviting someone to reinvent it in case of need. And the need can be easily anticipated.

jukkar replied:
I just want to avoid exposing functions if there is no real need for them. It is a pain to decprecate them if we find out we need to tweak the API or remove them.

pfalcon replied:
> I just want to avoid exposing functions if there is no real need for them.

While this PR is titled as SNTP-related, one of the subtasks was to start laying out helper API for apps (including POSIX apps) to reuse, instead of coding it again and again. So I did spend mental effort on factoring out that func to be (re)usable by other parts. That's why I'm concerned that this effort goes to /dev/null by request to static'ize it.

> It is a pain to decprecate them if we find out we need to tweak the API or remove them.

So again, that func just does a useful thing on its own (parses address in format host[:host]), and intended for reuse, so I don't see it being under a reasonable risk of deprecation. (I'd be more concerned with visibly adhoc codes like https://github.com/zephyrproject-rtos/zephyr/pull/15636 - that doesn't implement general enough API, so would be subject of long incremental tweaks and partial deprecation, or will need to be replaced with a new more generic impl).


That said, if whole this discussion doesn't convince you, and you'd rather see "static" now, hopefully removed later, I'd rather move on with the matter.

jukkar replied:
This is not that big of an issue. I just have been burned before because of introducing some functions that were "useful" too, but at the end they were really not useful after all.

pfalcon commented on subsys/net/lib/utils/addr_utils.c:
```diff
@@ -0,0 +1,60 @@
+	return NULL;
+}
+
+int net_getaddrinfo_addr_str(const char *addr_str, const char *def_port,
```
> As the functionality is almost identical to getaddrinfo(), why not call it net_getaddrinfo_str() as the addr looks redundant information.

The idea was to provide enough information and differentiation from raw getaddrinfo(), while keep it reasonably short. In that regard, as string in format `host[:port]` is called "address string", abbreviated `addr_str`, and that infix is used consistently in both function names, net_addr_str_find_port() and net_getaddrinfo_addr_str().

Even more selective name, like `addr_str_w_port`, would be better, but would lead to long function name. OTOH, just using "str" isn't selective enough IMHO.

pfalcon commented on subsys/net/lib/utils/addr_utils.c:
```diff
@@ -0,0 +1,60 @@
+			     const struct addrinfo *hints,
+			     struct addrinfo **res)
+{
+	const char *port = net_addr_str_find_port(addr_str);
```
> If we are calling this from user app, then we need to have more null checks for the input parameters.

Well, why? What's so special about NULL? What we'd be interested in catching *any* invalid value, but we can't, because user can write `char *foo` and get uninitiated var with value from stack. In this regard, NULL is just one of billions invalid value. If user passes NULL, the app will crush with a nice stacktrace (if built with suitable options), and we check for NULL, it doesn't crash there, but crashes elsewhere, or just silently misbehaves, in which regard checking for NULL (out of all other possible invalid values) is more trouble than boon.

d3zd3z replied:
Also, isn't checking argument one of those things the "Zephyr way" goes against? From a security perspective, this isn't really the best place to be checking values, and the ones that matter are those that come from external to the system.

jukkar replied:
As this function is to be callable from user space, it makes sense to validate the input.

pfalcon replied:
> As this function is to be callable from user space, it makes sense to validate the input.

It makes sense, but we can't really do that, as explained above, just can pretend to do that. All those NULL checks just add code size, but surprisingly, efficiency still seems to be a concern for Zephyr:  https://github.com/zephyrproject-rtos/zephyr/pull/15969.

 So, I'll add it here, but we'll end up needing to add CONFIG_DISABLE_NULL_CHECKS to disable all those useless checks sprinkled around and get the code size back.

pfalcon replied:
> As this function is to be callable from user space

And this argument... These functions are themselves userspace, they are just application convenience functions, factored out for reuse. It's not like that they're kernel-level functions, params to which should be validated (much more thoroughly than NULL checks).

Anyway, added NULL check.

pfalcon commented:
@jukkar: Thanks for the review. I addressed some points, and explained why other points were done as they were, which hopefully makes sense.

tgorochowik commented on subsys/net/lib/sntp/sntp_simple.c:
```diff
@@ -0,0 +1,44 @@
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_DGRAM;
+	hints.ai_protocol = 0;
+	res = net_getaddrinfo_addr_str(server, "123", &hints, &addr);
```
Shouldn't there be a define for the default port? Or at least a comment saying what it is?

pfalcon replied:
123 is the standard (S)NTP port per the RFC. Added comment on that.

tgorochowik commented on include/net/socketutils.h:
```diff
@@ -0,0 +1,35 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
```
Redundant newline

pfalcon replied:
Fixed

pfalcon commented:
Pushed new version addressing latest comments.
